include/nlohmann/detail/iterators/internal_iterator.hpp

```diff
@@ -1,25 +1,279 @@
 #pragma once
 
+#include <algorithm>
+#include <cstdint>
+#include <type_traits>
+#include <utility>
+
 #include <nlohmann/detail/iterators/primitive_iterator.hpp>
+#include <nlohmann/detail/macro_scope.hpp>
+#include <nlohmann/detail/meta/memory.hpp>
 
 namespace nlohmann
 {
 namespace detail
 {
+
+template<class BasicJsonType>
+class internal_iterator_base
+{
+  public:
+    using object_iterator_t = typename BasicJsonType::object_t::iterator;
+    using array_iterator_t = typename BasicJsonType::array_t::iterator;
+
+    object_iterator_t& object_iterator()
+    {
+        return *reinterpret_cast<object_iterator_t*>(&m_data);
+    }
+    const object_iterator_t& object_iterator() const
+    {
+        return *reinterpret_cast<const object_iterator_t*>(&m_data);
+    }
+
+    array_iterator_t& array_iterator()
+    {
+        return *reinterpret_cast<array_iterator_t*>(&m_data);
+    }
+    const array_iterator_t& array_iterator() const
+    {
+        return *reinterpret_cast<const array_iterator_t*>(&m_data);
+    }
+
+    primitive_iterator_t& primitive_iterator()
+    {
+        return *reinterpret_cast<primitive_iterator_t*>(&m_data);
+    }
+    const primitive_iterator_t& primitive_iterator() const
+    {
+        return *reinterpret_cast<const primitive_iterator_t*>(&m_data);
+    }
+
+  private:
+    struct
+    {
+#ifdef JSON_HAS_CPP_14
+        // NOLINTNEXTLINE
+        alignas(std::max({alignof(object_iterator_t), alignof(array_iterator_t), alignof(primitive_iterator_t)})) std::uint8_t m_t[std::max({sizeof(object_iterator_t), sizeof(array_iterator_t), sizeof(primitive_iterator_t)})];
+#else
+        // Below code is horrible, but it's the only way to make gcc 4.8 compile
+        alignas(alignof(object_iterator_t) < alignof(array_iterator_t) ? (alignof(array_iterator_t) < alignof(primitive_iterator_t)
+                ? alignof(primitive_iterator_t)
+                : alignof(array_iterator_t))
+                : (alignof(object_iterator_t) < alignof(primitive_iterator_t)
+                   ? alignof(primitive_iterator_t)
+                   // NOLINTNEXTLINE
+                   : alignof(object_iterator_t))) std::uint8_t m_t[sizeof(object_iterator_t) < sizeof(array_iterator_t) ? (sizeof(array_iterator_t) < sizeof(primitive_iterator_t)
+                                                     ? sizeof(primitive_iterator_t)
+                                                     : sizeof(array_iterator_t))
+                                                     : (sizeof(object_iterator_t) < sizeof(primitive_iterator_t)
+                                                        ? sizeof(primitive_iterator_t)
+                                                        : sizeof(object_iterator_t))];
+#endif
+    } m_data {};
+};
```

Comment:
Horrible, indeed. I think I have a better solution. Will get back to you when I have more time.

Reply:
This should work:

https://godbolt.org/z/MGhhEsd3G

```cpp
namespace detail
{

template<std::size_t... Is>
struct max;

template<>
struct max<> {
    using type = std::integral_constant<std::size_t, 0>;
};

template<std::size_t I, std::size_t... Rest>
struct max<I, Rest...> {
    using rest_type = typename max<Rest...>::type;
    using type = typename std::conditional<
        (I > rest_type::value),
        std::integral_constant<std::size_t, I>,
        rest_type>::type;
};

template<typename... Ts>
using max_alignof = typename max<alignof(Ts)...>::type;

template<typename... Ts>
using max_sizeof = typename max<sizeof(Ts)...>::type;

}

constexpr std::size_t data_align =
    detail::max_alignof<int, double, char>::value;
constexpr std::size_t data_size =
    detail::max_sizeof<int, double, char>::value;

alignas(data_align) std::uint8_t data[data_size];
```

Maybe add a new file `meta/algorithm.hpp` or similar.